File: src/vscode-windhawk/src/coreClient/parseModSource.ts
import ModSource from '../services/modSource';
import { ParsedModSource } from './contract';

// Shared implementation of the contract's parseModSource command: parse the
// three sections independently so one malformed block doesn't hide the
// others. Used by the in-process backend (against its ModSource instance)
// and by parseModSourceStandalone below.
export function parseModSourceWith(
	modSource: ModSource,
	source: string,
	language: string,
): ParsedModSource {
	const result: ParsedModSource = {
		metadata: null,
		readme: null,
		initialSettings: null,
		errors: {},
	};

	try {
		result.metadata = modSource.extractMetadata(source, language);
	} catch (e) {
		result.errors.metadata = e instanceof Error ? e.message : String(e);
	}

	try {
		result.readme = modSource.extractReadme(source);
	} catch (e) {
		result.errors.readme = e instanceof Error ? e.message : String(e);
	}

	try {
		result.initialSettings = modSource.extractInitialSettings(source, language);
	} catch (e) {
		result.errors.initialSettings = e instanceof Error ? e.message : String(e);
	}

	return result;
}

// Session-free variant for callers that have no Windhawk environment at all
// (the CLI's `source meta`, which works without an app root). parseModSource is
// a pure helper, so this is the same code path the session command takes; the
// ModSource extract methods use no instance state, making an empty appDataPath
// safe.
export function parseModSourceStandalone(source: string, language: string): ParsedModSource {
	return parseModSourceWith(new ModSource(''), source, language);
}

File: src/windhawk-core/core/src/commands/parse_mod_source.rs
//! `parseModSource`: parse metadata, readme, and initial settings out of mod
//! source code. Each section parses independently so one malformed block
//! doesn't hide the others; a failed section yields a per-section error string,
//! never a command failure.

use serde_json::Value;
use windhawk_core_domain as domain;
use windhawk_core_protocol::{
    AppendToModIdAndNameParams, ParseModSourceParams, ParsedModSource, ParsedModSourceErrors,
};

use crate::convert::{metadata_to_protocol, settings_to_protocol};
use crate::dispatch::decode_params;
use crate::error::CoreError;

// Pure, session-free handlers (`Handler::Stateless`): they read no session
// state, so they take only the request params.
pub fn run(params: Value) -> Result<Value, CoreError> {
    let params: ParseModSourceParams = decode_params("parseModSource", params)?;

    let mut result = ParsedModSource {
        metadata: None,
        readme: None,
        initial_settings: None,
        errors: ParsedModSourceErrors::default(),
    };

    match domain::extract_metadata(&params.source, &params.language) {
        Ok(metadata) => result.metadata = Some(metadata_to_protocol(metadata)),
        Err(e) => result.errors.metadata = Some(e.to_string()),
    }

    // Readme extraction has no failure mode: an absent or malformed block
    // is null, exactly like the TS implementation.
    result.readme = domain::extract_readme(&params.source);

    match domain::extract_initial_settings(&params.source, &params.language) {
        Ok(settings) => result.initial_settings = settings.map(settings_to_protocol),
        Err(e) => result.errors.initial_settings = Some(e.to_string()),
    }

    serde_json::to_value(&result)
        .map_err(|e| CoreError::internal(format!("parseModSource result serialization: {e}")))
}

/// `appendToModIdAndName`: a pure source transform (the new-mod / fork flows),
/// dispatch-direct into `domain` like `parseModSource`. Returns the transformed
/// source as a bare JSON string.
pub fn append_mod_id_and_name(params: Value) -> Result<Value, CoreError> {
    let params: AppendToModIdAndNameParams = decode_params("appendToModIdAndName", params)?;
    let transformed = domain::append_to_id_and_name(
        &params.source,
        params.append_to_id.as_deref(),
        params.append_to_name.as_deref(),
    );
    Ok(Value::String(transformed))
}
